File: packages/depz-sensor-sdk-csharp/src/Depz.Sensor/Transport/PacketParser.cs
namespace Depz.Sensor.Transport;

/// <summary>
/// Incremental frame parser. Feed arbitrary byte chunks; get events.
///
/// Event order is invariant to chunking (contract 01 §5) except <see cref="Trash"/>
/// event boundaries — concatenate Trash data when comparing streams. Byte-exact
/// with the reference Python <c>PacketParser</c>, including the
/// empty-payload-no-CRC rule (ERRATA E6) and single-byte resync on a corrupt
/// header.
/// </summary>
public sealed class PacketParser
{
    private readonly List<byte> _buf = new();

    public int Packets { get; private set; }
    public int CrcErrors { get; private set; }
    public int HeaderErrors { get; private set; }
    public int TrashBytes { get; private set; }

    /// <summary>Current unconsumed bytes (residue).</summary>
    public byte[] Residue() => _buf.ToArray();

    public IReadOnlyList<ParserEvent> Feed(ReadOnlySpan<byte> data)
    {
        foreach (byte b in data)
            _buf.Add(b);
        var outEvents = new List<ParserEvent>();
        while (true)
        {
            ParserEvent? ev = ParseOne();
            if (ev is null)
                break;
            outEvents.Add(ev);
        }
        return outEvents;
    }

    private Trash EmitTrash(int count)
    {
        var data = new byte[count];
        _buf.CopyTo(0, data, 0, count);
        _buf.RemoveRange(0, count);
        TrashBytes += count;
        return new Trash(data);
    }

    private int FindMagic()
    {
        for (int i = 0; i + 1 < _buf.Count; i++)
        {
            if (_buf[i] == Framing.Magic[0] && _buf[i + 1] == Framing.Magic[1])
                return i;
        }
        return -1;
    }

    private ParserEvent? ParseOne()
    {
        int pos = FindMagic();
        if (pos == -1)
        {
            // Keep the last byte: it may be a split 0xA5.
            if (_buf.Count > 1)
                return EmitTrash(_buf.Count - 1);
            return null;
        }
        if (pos > 0)
            return EmitTrash(pos);
        if (_buf.Count < Framing.HeaderSize)
            return null;

        int dataSize = _buf[2] | (_buf[3] << 8);
        int payloadSize = dataSize & Framing.MaxPayload;
        var crcType = (CrcType)((dataSize >> 14) & 0x03);

        byte[] hdrInput = { _buf[2], _buf[3], _buf[4], _buf[5] };
        if (Crc.Crc8Maxim(hdrInput) != _buf[6])
        {
            // Header corrupt: advance one byte past the magic start and let the
            // magic hunt resync (firmware: rb_skip(off + 1)).
            HeaderErrors++;
            return EmitTrash(1);
        }

        // ERRATA E6: empty payload never carries CRC bytes.
        int crcSize = payloadSize > 0 ? crcType.Size() : 0;
        int total = Framing.HeaderSize + payloadSize + crcSize;
        if (_buf.Count < total)
            return null;

        int cmd = _buf[4];
        int seq = _buf[5];
        var payload = new byte[payloadSize];
        _buf.CopyTo(Framing.HeaderSize, payload, 0, payloadSize);
        var trailer = new byte[crcSize];
        _buf.CopyTo(Framing.HeaderSize + payloadSize, trailer, 0, crcSize);
        _buf.RemoveRange(0, total);

        if (crcSize > 0 && !Framing.PayloadCrcBytes(crcType, payload).AsSpan().SequenceEqual(trailer))
        {
            CrcErrors++;
            return new CrcError(cmd, seq);
        }
        Packets++;
        return new Packet(cmd, seq, payload);
    }
}

File: packages/depz-sensor-sdk-java/src/main/java/ai/depz/sensor/transport/PacketParser.java
package ai.depz.sensor.transport;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Incremental frame parser. Feed arbitrary byte chunks; get events.
 *
 * <p>Event order is invariant to chunking (contract 01 §5) except Trash event
 * boundaries — concatenate Trash data when comparing streams. Byte-exact with
 * the reference Python {@code PacketParser}, including the empty-payload-no-CRC
 * rule (ERRATA E6) and single-byte resync on a corrupt header.
 */
public final class PacketParser {
    private byte[] buf = new byte[64];
    private int len = 0;

    public int packets = 0;
    public int crcErrors = 0;
    public int headerErrors = 0;
    public int trashBytes = 0;

    private void ensureCapacity(int extra) {
        if (len + extra > buf.length) {
            int cap = buf.length;
            while (cap < len + extra) {
                cap *= 2;
            }
            buf = Arrays.copyOf(buf, cap);
        }
    }

    private void delFront(int n) {
        System.arraycopy(buf, n, buf, 0, len - n);
        len -= n;
    }

    private int findMagic() {
        for (int i = 0; i + 1 < len; i++) {
            if (buf[i] == Framing.MAGIC[0] && buf[i + 1] == Framing.MAGIC[1]) {
                return i;
            }
        }
        return -1;
    }

    /** Current unconsumed bytes (residue). */
    public byte[] residue() {
        return Arrays.copyOf(buf, len);
    }

    public List<Event> feed(byte[] data) {
        ensureCapacity(data.length);
        System.arraycopy(data, 0, buf, len, data.length);
        len += data.length;
        List<Event> out = new ArrayList<>();
        while (true) {
            Event ev = parseOne();
            if (ev == null) {
                break;
            }
            out.add(ev);
        }
        return out;
    }

    private Trash emitTrash(int count) {
        byte[] data = Arrays.copyOf(buf, count);
        delFront(count);
        trashBytes += count;
        return new Trash(data);
    }

    private static int u16le(byte[] b, int off) {
        return (b[off] & 0xFF) | ((b[off + 1] & 0xFF) << 8);
    }

    private Event parseOne() {
        int pos = findMagic();
        if (pos == -1) {
            // Keep the last byte: it may be a split 0xA5.
            if (len > 1) {
                return emitTrash(len - 1);
            }
            return null;
        }
        if (pos > 0) {
            return emitTrash(pos);
        }
        if (len < Framing.HEADER_SIZE) {
            return null;
        }
        int dataSize = u16le(buf, 2);
        int payloadSize = dataSize & Framing.MAX_PAYLOAD;
        CrcType crcType = CrcType.fromValue((dataSize >>> 14) & 0x03);
        byte[] hdrInput = {buf[2], buf[3], buf[4], buf[5]};
        if (Crc.crc8Maxim(hdrInput) != (buf[6] & 0xFF)) {
            // Header corrupt: advance one byte past the magic start and let the
            // magic hunt resync (firmware: rb_skip(off + 1)).
            headerErrors++;
            return emitTrash(1);
        }
        // ERRATA E6: empty payload never carries CRC bytes.
        int crcSize = payloadSize > 0 ? crcType.size : 0;
        int total = Framing.HEADER_SIZE + payloadSize + crcSize;
        if (len < total) {
            return null;
        }
        int cmd = buf[4] & 0xFF;
        int seq = buf[5] & 0xFF;
        byte[] payload = Arrays.copyOfRange(buf, Framing.HEADER_SIZE, Framing.HEADER_SIZE + payloadSize);
        byte[] trailer = Arrays.copyOfRange(buf, Framing.HEADER_SIZE + payloadSize, total);
        delFront(total);
        if (crcSize > 0 && !Arrays.equals(Framing.payloadCrcBytes(crcType, payload), trailer)) {
            crcErrors++;
            return new CrcError(cmd, seq);
        }
        packets++;
        return new Packet(cmd, seq, payload);
    }
}
